#BiocManager::install("tximport")
library("tximport")

acc_list <- read.delim("raw_data_names_short.txt", header = FALSE)
files <- file.path("quants", acc_list$V1,"quant.sf")
names(files) <- as.character(acc_list$V1)

# prepare tx2gene dataframe
tx2gene <- read.delim(file.path("reference", "Niben101_annotation.transcripts.fa", "tx2gene.txt"), stringsAsFactors = FALSE)
#The tximport arguments varReduce and dropInfReps can be used to summarize the inferential replicates into a single variance per transcript/gene and per sample, or to not import inferential replicates, respectively.
#tximport 参数 varReduce 和 dropInfReps 可用于将推论复制总结为每个转录本/基因和每个样本的单个方差，或者分别不导入推论复制。
txi.salmon <- tximport(files, type = "salmon", tx2gene = tx2gene, dropInfReps=TRUE)
str(txi.salmon)#12 samples and 56883genes were reported

head(txi.salmon$abundance)#It typically contains TPM (Transcripts Per Million) values.
#TPM is a normalized unit of expression that accounts for:
#Gene/transcript length (longer genes get more reads just by chance) and Sequencing depth (number of total reads in a sample)
#Used for Descriptive analysis, visualization,within-sample comparisons

head(txi.salmon$counts)#use counts for DE analysis. 
#This is the matrix of estimated raw counts, summed to the gene level (using tx2gene).
#These are adjusted from transcript-level estimates, but not yet normalized.
#They are suitable input for DESeq2 or edgeR, which handle normalization internally 
#(DESeq2 uses a method called "median of ratios" to normalize for:Sequencing depth (library size differences between samples) and RNA composition bias (e.g., a few highly expressed genes skewing total counts)).
#Used forDESeq2 / differential expression


pdf("histogram_tpm.pdf")
hist(log2(txi.salmon$abundance),
     main = "Distribution of TPM values",
     xlab = "log2 TPM",
     ylab = "nr. of genes across all samples",
     col = "#e5f5e0",
     border = "#31a354")
dev.off()

# write the gene level TPM values into a file
##This is gene level counts generated by salmon
write.table(txi.salmon$counts, file = "salmon_counts_counts.txt", sep = "\t", row.names = TRUE)

# Salmon puts out tpm values normalized for transcript length and library size
# you can use an R script to analyze expression:
# "get_gene_level_quant.R" summarizes reads for every gene instead of transcripts